logstash-core/src/test/java/org/logstash/execution/ShutdownWatcherExtTest.java

```diff
@@ -20,69 +20,122 @@
+
+import org.apache.logging.log4j.junit.LoggerContextRule;
+import org.apache.logging.log4j.test.appender.ListAppender;
 import org.assertj.core.api.Assertions;
 import org.jruby.RubySystemExit;
 import org.jruby.exceptions.RaiseException;
 import org.jruby.runtime.ThreadContext;
 import org.jruby.runtime.builtin.IRubyObject;
+import org.junit.Before;
+import org.junit.ClassRule;
 import org.junit.Test;
 import org.logstash.RubyTestBase;
 import org.logstash.RubyUtil;
 
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
 /**
  * Tests for {@link ShutdownWatcherExt}.
  */
 @NotThreadSafe
 public final class ShutdownWatcherExtTest extends RubyTestBase {
 
+    private static final String CONFIG = "log4j2-test1.xml";
+    private ListAppender appender;
+
+    @ClassRule
+    public static LoggerContextRule CTX = new LoggerContextRule(CONFIG);
+
+    @Before
+    public void setup() {
+        appender = CTX.getListAppender("EventLogger").clear();
+    }
+
+    public List<CharSequence> pipeline = Arrays.asList(
+            "pipeline = Object.new",
+            "reporter = Object.new",
+            "snapshot = Object.new",
+            "inflight_count = java.util.concurrent.atomic.AtomicInteger.new",
+            "snapshot.define_singleton_method(:inflight_count) do",
+            "inflight_count.increment_and_get + 1",
+            "end",
+            "threads = {}",
+            "snapshot.define_singleton_method(:stalling_threads) do",
+            "threads",
+            "end",
+            "snapshot.define_singleton_method(:to_s) do",
+            "\"inflight_count=>\" + inflight_count.get.to_s + \", stalling_threads_info=>{...}\"",
+            "end",
+            "reporter.define_singleton_method(:snapshot) do",
+            "snapshot",
+            "end",
+            "pipeline.define_singleton_method(:thread) do",
+            "Thread.current",
+            "end",
+            "pipeline.define_singleton_method(:finished_execution?) do",
+            "false",
+            "end",
+            "pipeline.define_singleton_method(:reporter) do",
+            "reporter",
+            "end",
+            "pipeline.define_singleton_method(:worker_threads_draining?) do",
+            "false",
+            "end",
+            "pipeline"
+    );
+
     @Test
-    public void testShouldForceShutdown() throws InterruptedException {
+    public void pipelineWithUnsafeShutdownShouldForceShutdown() throws InterruptedException {
+        watcherShutdownStallingPipeline(pipeline);
+
+        // non drain pipeline should print stall msg
+        boolean printStalling = appender.getMessages().stream().anyMatch((msg) -> msg.contains("stalling"));
+        assertTrue(printStalling);
+    }
+
+
+    @Test
+    public void pipelineWithDrainShouldNotPrintStallMsg() throws InterruptedException {
+        // set drain to true
+        List<CharSequence> drainPipeline = new ArrayList<>(pipeline);
```

Comment:
I think this part, the stream to calculate the index that we want to change, is little bit over complicated.

I think we could simplify in 2 ways:
- we know upfront that  the line we want to change is 28th, so change it directly without computing the index. As a safe check that's the correct one, we could check the previous contains the "draining?"
```java
assertThat(drainPipeline.getAt(27 - 1), containsString("draing?"));
drainPipeline.set(27, "true");
```
- or we can use a template for the RubyString, at line 28th instead of `false`  we set "%{return_value_placeholer}" and write a method to replace it.

Reply:
I was initially hard code 27 instead of searching the index, but thinking of the flexibility of extending the pipeline and not to trap the next one who wants to modify the pipeline, setting a fixed index doesn't seem to be a good approach. I prefer to keep it as is.

Reply:
Do you think that moving to a template and apply a replacement strategy would be better?
Reading the stream operation to understand that what it does is search the right position to replace "false" with "true" is not immediate.

Reply:
your suggestion is to use `%{value_placeholder}` instead of setting a default value. For every usage, search for the index of placeholder and then set true/ false. Or do you mean using erb template? 
I think both improve readability. The former is quite similar to the current setup. Setting the default value is also not too bad. I am open to make the change

Reply:
I wouldn't user `erb` template from java code. I would simply put a file in the `test/resources` with the placeholder, and write a utility method, used by the test, to replace the placeholder.